base/special/gamma.jl

```diff
@@ -496,20 +498,71 @@ for (f,T) in ((:f32,Float32),(:f16,Float16))
     end
 end
 
-zeta(s::Integer, z::Number) = zeta(Int(s), f64(z))
-zeta(s::Number, z::Number) = zeta(f64(s), f64(z))
-for f in (:digamma, :trigamma)
+
+function zeta(s::Integer, z::Number) 
+    x = float(z)
+    t = Int(s)
+    if typeof(x) === typeof(z) && typeof(t) === typeof(s)
+        throw MethodError(zeta,(s,t))
+    end
+    oftype(x, zeta(t, x))
+end
+
+function zeta(s::Number, z::Number) 
+    x = float(z)
+    t = float(s)
+    if typeof(x) === typeof(z) && typeof(t) === typeof(s)
+        throw MethodError(zeta,(s,t))
+    end
+    oftype(x, zeta(t, x))
+end
+
+
+function polygamma(m::Integer, z::Number)
+    x = float(z)
+    typeof(x) == typeof(z) && throw(MethodError(polygamma, (m,z))
+    oftype(x,polygamma(m, x))
+end
+
+
+for f in (:digamma, :trigamma, :zeta, :eta, invdigamma)
     @eval begin
-        $f(z::Number) = $f(f64(z))
+        $f(z::Base.BitInteger) = $f(Float64(z))
+        $f(z::Float32) = Float32($f(Float64(z)))
+        $f(z::Float16) = Float16($f(Float64(z)))
+
+        function $f(z::Number)
+            x = float(z)
+            typeof(x) == typeof(z) && throw(MethodError($f, (z,)))
+            oftype(x, $f(x))
+        end
     end
 end
-polygamma(m::Integer, z::Number) = polygamma(m, f64(z))
 
-# Inverse digamma function:
-# Implementation of fixed point algorithm described in
-#  "Estimating a Dirichlet distribution" by Thomas P. Minka, 2000
+for f in (:zeta, :eta)
+    @eval begin
+        $f{T<:Union{Base.BitInteger,Float32,Float16}}(x::t) = oftype(float(x), $f(Complex128(z)))
```

Comment:
`Complex128` is a type alias for `Complex{Float64}`.
I guess cos it is made of 2 Float64s
It is odd that it is converting a real to a complex though.

Looking at it, it is because that line doesn't even compile. (I got distracted when writing this function i guess).
I know I've fixed it already in my local copy.
`x::t` is `x::Complex{T}`
Or probably `x::Complex{Union{Base.BitInteger,Float32,Float16}}`
Can't remember